cardano-constitution/test/Cardano/Constitution/Validator/Data/GoldenTests/unsorted.pir.golden

```diff
@@ -423,20 +423,12 @@ program
               !args : list data = sndPair {integer} {list data} tup
             in
             case
-              (all dead. Maybe a)
-              (equalsInteger 1 index)
-              [ (/\dead ->
-                   case
-                     (all dead. Maybe a)
-                     (equalsInteger 0 index)
-                     [ (/\dead -> error {Maybe a})
-                     , (/\dead ->
-                          Just
-                            {a}
-                            (`$dUnsafeFromData` (headList {data} args))) ]
-                     {all dead. dead})
-              , (/\dead -> Nothing {a}) ]
-              {all dead. dead}
+              (list data -> Maybe a)
+              index
+              [ (\(ds : list data) ->
+                   Just {a} (`$dUnsafeFromData` (headList {data} ds)))
+              , (\(ds : list data) -> Nothing {a}) ]
```

Comment:
Would it make sense to pull `\(ds : list data) -> <...>` out of `case`, given that it appears in both the branches? If only to save on size. Not as a general optimization, because it's semantics changing, since `case` may fail, but can we just change the relevant Plinth code or whatever?

That'd also play better with inlining, case-of-case and whatnot.

Reply:
This is quite a bit of more work. What we generate using TH is essentially a case matching with a ton of view patterns attached. Which is useful for the "safe" fromData, but not so much for unsafe one. 

Unfortunately, I'd had to redo pretty big chunk of TH to do this since many part of it is very intertwined. I prefer to have this as a separate PR, if we decide that we want this.